spec/entropy.md

```diff
@@ -0,0 +1,168 @@
+
+# Global structure
+
+```
+Stream ::= Headers Prelude? Content? Main Footers?
+```
+
+# Headers
+
+Headers identify the file, the version of the format used and the shared dictionary used, if any.
+
+The shared dictionary provides:
+
+- a family of context-based prediction tables;
+- well-known literal strings;
+- well-known floating-point numbers;
+- well-known unsigned long numbers;
+- well-known property keys;
+- well-known identifier names;
+- well-known list lengths.
+
+```
+Headers ::= MagicNumber FormatVersionNumber LinkToSharedDictionary
+```
+
+## Magic number
+
+```
+MagicNumber ::= "\x89BJS\r\n\0\n"
+```
+
+
+## Version number
+
+Not specified/implemented yet.
+
+```
+VersionNumber ::= TBD
+```
+
+## Link to shared dictionary
+
+Not specified/implemented yet.
+
+```
+LinkToSharedDictionary ::= TBD
+```
+
+
+# Prelude
+
+The prelude extends the shared dictionary with additional strings, numbers, keys, names, ...
+
+```
+Prelude ::= "[[prelude]]" PreludeDictionary*
+PreludeDictionary ::= "[floats]" PreludeStream<f64?>
+                   |  "[unsigned_longs]" PreludeStream<u32?>
+                   |  "[list_lengths]" PreludeStream<u32?>
+                   |  "[property_keys]" PreludeStream<String>
+                   |  "[identifier_names]" PreludeStream<String>
+                   |  "[string_literals]" PreludeStream<String>
+                   |  "[property_keys_len]" PreludeStream<StringLen>
+                   |  "[identifier_names_len]" PreludeStream<StringLen>
+                   |  "[string_literals_len]" PreludeStream<StringLen>
+```
+
+If the same prelude name (`"[floats]"`, `"[unsigned_longs]"`, etc.) appears more than once, this
+is a syntax error.
+
+If `["property_keys"]` exists and is not empty, `'["property_keys_len"]` must exist and be non-empty,
+otherwise, this is a syntax error.
+
+If `["identifier_names"]` exists and is not empty, `'["identifier_names"]` must exist and be non-empty,
+otherwise, this is a syntax error.
+
+If `["string_literals"]` exists and is not empty, `'["string_literals_len"]` must exist and be non-empty,
+otherwise, this is a syntax error.
+
+## Internal compression
+
+All streams have the same structure
+
+```
+PreludeStream<T> ::= CompressionFormat ";" ByteLen CompressedByteStream<T>
+CompressionFormat ::= "brotli" // The only compression format supported so far.
+ByteLen := Varu32
+```
+
+Where:
+
+- the `CompressionFormat` is the name of the compression to use to decompress the `CompressedByteStream`;
+- the `ByteLen` is the number of bytes in the `CompressedByteStream`;
+
+Let us call `ExpandedByteStream<T>` the result of decompressing the `ByteLen` bytes of
+a `CompressedByteStream<T>` with the specified `CompressionFormat`.
+
+## Stream interpretation
+
+### Integers
+
+```
+ExpandedByteStream<u32?> ::= ExpandedByteStreamItem<u32?>*
+ExpandedByteStreamItem<u32?> ::= Varu32         // Some(value)
+                              | Varu32Invalid1  // None
```

Comment:
one more space for indent

Reply:
Gasp, I learnt to center the `|` under the `::=` :)